sklearn/tests/test_impute.py

```diff
@@ -210,6 +230,161 @@ def test_imputation_most_frequent():
 
 
+def test_imputation_most_frequent_objects():
+    # Test imputation using the most-frequent strategy.
+    for marker in (None, np.nan, "NAN", "", 0):
+        X = np.array([
+            [marker, marker, "a", "f"],
+            [marker, "c", marker, "d"],
+            [marker, "b", "d", marker],
+            [marker, "c", "d", "h"],
+        ], dtype=object)
+
+        X_true = np.array([
+            ["c", "a", "f"],
+            ["c", "d", "d"],
+            ["b", "d", "d"],
+            ["c", "d", "h"],
+        ], dtype=object)
+
+        imputer = SimpleImputer(missing_values=marker,
+                                strategy="most_frequent")
+        X_trans = imputer.fit(X).transform(X)
+
+        assert_array_equal(X_trans, X_true)
+
+
+def test_imputation_most_frequent_pandas():
+    # Test imputation using the most frequent strategy on pandas df
+    pd = pytest.importorskip("pandas")
+
+    f = io.StringIO(u"Cat1,Cat2,Cat3,Cat4\n"
+                    ",i,x,\n"
+                    "a,,y,\n"
+                    "a,j,,\n"
+                    "b,j,x,")
+
+    for dtype in (object, "category"):
+        df = pd.read_csv(f, dtype=dtype)
+        f.seek(0)
+
+        X_true = np.array([
+            ["a", "i", "x"],
+            ["a", "j", "y"],
+            ["a", "j", "x"],
+            ["b", "j", "x"]
+        ], dtype=object)
+
+        imputer = SimpleImputer(strategy="most_frequent")
+        X_trans = imputer.fit(df).transform(df)
+
+        assert_array_equal(X_trans, X_true)
+
+
+def test_imputation_constant_integer():
+    # Test imputation using the constant strategy on integers
+    X = np.array([
+        [-1, 2, 3, -1],
+        [4, -1, 5, -1],
+        [6, 7, -1, -1],
+        [8, 9, 0, -1]
+    ])
+
+    X_true = np.array([
+        [0, 2, 3, 0],
+        [4, 0, 5, 0],
+        [6, 7, 0, 0],
+        [8, 9, 0, 0]
+    ])
+
+    imputer = SimpleImputer(missing_values=-1, strategy="constant",
+                            fill_value=0)
+    X_trans = imputer.fit(X).transform(X)
+
+    assert_array_equal(X_trans, X_true)
+
+
+def test_imputation_constant_float():
+    # Test imputation using the constant strategy on floats
+    for format in ["csr", "array"]:
+        X = np.array([
+            [np.nan, 1.1, 2.2, np.nan],
+            [3.3, np.nan, 4.4, np.nan],
+            [5.5, 6.6, np.nan, np.nan],
+            [7.7, 8.8, 9.9, np.nan]
+        ])
+
+        X = sparse.csr_matrix(X) if format == "csr" else X
+
+        X_true = np.array([
+            [0, 1.1, 2.2, 0],
+            [3.3, 0, 4.4, 0],
+            [5.5, 6.6, 0, 0],
+            [7.7, 8.8, 9.9, 0]
+        ])
+
+        X_true = sparse.csr_matrix(X_true) if format == "csr" else X_true
+
+        imputer = SimpleImputer(strategy="constant", fill_value=0)
+        X_trans = imputer.fit(X).transform(X)
+
+        if format == "csr":
+            assert_allclose(X_trans.toarray(), X_true.toarray())
```

Comment:
you have a function `assert_allclose_dense_sparse` in testing which can support both case without distinction